docs/templates/feature.md

```diff
@@ -51,27 +50,77 @@ Potential *offline* types are:
-    providing a schema manually, the feature store will infer the schema with feature
-    types from the DataFrame.
-
 Potential *online* types are:
 
 ```SQL
 "None", "INT(11)", "TINYINT(1)", "SMALLINT(5)", "MEDIUMINT(7)", "BIGINT(20)",
 "FLOAT", "DOUBLE", "DECIMAL", "DATE", "DATETIME", "TIMESTAMP", "TIME", "YEAR",
-"CHAR", "VARCHAR(25)", "VARCHAR(125)", "VARCHAR(225)", "VARCHAR(500)",
-"VARCHAR(1000)", "VARCHAR(2000)", "VARCHAR(5000)", "VARCHAR(10000)", "BINARY",
-"VARBINARY(100)", "VARBINARY(500)", "VARBINARY(1000)", "BLOB", "TEXT",
-"TINYBLOB", "TINYTEXT", "MEDIUMBLOB", "MEDIUMTEXT", "LONGBLOB", "LONGTEXT",
-"JSON"
+"CHAR", "VARCHAR(n)", "BINARY", "VARBINARY(n)", "BLOB", "TEXT", "TINYBLOB",
+"TINYTEXT", "MEDIUMBLOB", "MEDIUMTEXT", "LONGBLOB", "LONGTEXT", "JSON"
 ```
 
+#### Complex online types
+
+Additionally to the *online* types above, Hopsworks allows users to store complex types (e.g. *ARRAY<INT>*) on the online feature store.
+Hopsworks will take care of transparently serializing the complex features and storing them as *VARBINARY* on the online feature store. The serialization happens when calling the [save()](../api/feature_group_api/#save), [insert()](../api/feature_group_api/#insert) or [insert_stream()](../api/feature_group_api/#insert_stream) methods. The deserialization will be executed when calling the [get_serving_vector()](../api/training_dataset_api/#get_serving_vector) method to retrieve data from the online feature store.
+If users query directly the online feature store, for instance using the `fs.sql("SELECT ...", online=True)` statement, they will receive back a binary blob.
+
+On the feature store UI, the online feature type for complex features will be reported as *VARBINARY*.
+
+#### Online restrictions for primary key types:
+
+When a feature is being used as a primary key, certain types are not allowed. Examples of such types are *Float*, *Double*, *Date*, *Text*, *Blob* and *Complex Types*  (e.g. Array<>). Additionally the size of the sum of the primary key online types storage requirements should not exceed 3KB.
+
+### Type Inference
+
+The offline and online types for each feature are inferred automatically when calling the [save()](../api/feature_group_api/#save) method to create a feature group. The types will be inferred based on the types of the Spark or Pandas DataFrame.
+
+In the case of Spark DataFrame, the [Spark types](https://spark.apache.org/docs/latest/sql-ref-datatypes.html) will be converted in the equivalent Hive Metastore type and used as offline feature store type. If the feature group is online enabled, Hopsworks will then match the offline type to the corresponding online type. The matching is performed as following:
+
+* If the offline type is supported also on the online feature store (e.g. INT, FLOAT, DATE, TIMESTAMP), the online type will be equivalent to the offline type
+* If the offline type is *boolean*, the online type is going to be set as *tinyint*
```

Comment:
change all "online type" and "offline type" to "online data type" and "offline data type"